// @juit/qrcode/dist/encode.mjs
// encode.ts
var ALPHANUM = (function(s) {
  const res = {};
  for (let i = 0; i < s.length; i++) {
    res[s[i]] = i;
  }
  return res;
})("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:");
function pushBits(arr, n, value) {
  for (let bit = 1 << n - 1; bit; bit = bit >>> 1) {
    arr.push(!!(bit & value));
  }
  return arr;
}
function binaryEncode(data) {
  const len = data.length;
  const bits = [];
  for (let i = 0; i < len; i++) {
    pushBits(bits, 8, data[i]);
  }
  const d = pushBits([false, true, false, false], 16, len);
  const res = {
    data27: d.concat(bits)
  };
  res.data10 = res.data27;
  if (len < 256) {
    const d2 = pushBits([false, true, false, false], 8, len);
    res.data1 = d2.concat(bits);
  }
  return res;
}
function alphanumEncode(str) {
  const len = str.length;
  const bits = [];
  for (let i = 0; i < len; i += 2) {
    let b = 6;
    let n = ALPHANUM[str[i]];
    if (str[i + 1]) {
      b = 11;
      n = n * 45 + ALPHANUM[str[i + 1]];
    }
    pushBits(bits, b, n);
  }
  const d = pushBits([false, false, true, false], 13, len);
  const res = {
    data27: d.concat(bits)
  };
  if (len < 2048) {
    const d2 = pushBits([false, false, true, false], 11, len);
    res.data10 = d2.concat(bits);
  }
  if (len < 512) {
    const d2 = pushBits([false, false, true, false], 9, len);
    res.data1 = d2.concat(bits);
  }
  return res;
}
function numericEncode(str) {
  const len = str.length;
  const bits = [];
  for (let i = 0; i < len; i += 3) {
    const s = str.substring(i, i + 3);
    const b = Math.ceil(s.length * 10 / 3);
    pushBits(bits, b, parseInt(s, 10));
  }
  const d = pushBits([false, false, false, true], 14, len);
  const res = {
    data27: d.concat(bits)
  };
  if (len < 4096) {
    const d2 = pushBits([false, false, false, true], 12, len);
    res.data10 = d2.concat(bits);
  }
  if (len < 1024) {
    const d2 = pushBits([false, false, false, true], 10, len);
    res.data1 = d2.concat(bits);
  }
  return res;
}
function urlEncode(str) {
  const slash = str.indexOf("/", 8) + 1 || str.length;
  const res = encodeQrCodeMessage(str.slice(0, slash).toUpperCase(), false);
  if (slash >= str.length) return res;
  const path = encodeQrCodeMessage(str.slice(slash), false);
  res.data27 = res.data27.concat(path.data27);
  if (res.data10 && path.data10) {
    res.data10 = res.data10.concat(path.data10);
  }
  if (res.data1 && path.data1) {
    res.data1 = res.data1.concat(path.data1);
  }
  return res;
}
function encodeQrCodeMessage(message, url) {
  let data;
  if (typeof message === "string") {
    data = new TextEncoder().encode(message);
    if (/^[0-9]+$/.test(message)) {
      if (data.length > 7089) throw new Error(`Too much numeric data (len=${data.length})`);
      return numericEncode(message);
    }
    if (/^[0-9A-Z $%*+./:-]+$/.test(message)) {
      if (data.length > 4296) throw new Error(`Too much alphanumeric data (len=${data.length})`);
      return alphanumEncode(message);
    }
    if (url && /^https?:/i.test(message)) {
      return urlEncode(message);
    }
  } else {
    data = message;
  }
  if (data.length > 2953) throw new Error(`Too much binary data (len=${data.length})`);
  return binaryEncode(data);
}
export {
  encodeQrCodeMessage
};
