// @juit/qrcode/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// index.ts
var index_exports = {};
__export(index_exports, {
  generate: () => generate,
  generatePdf: () => import_pdf2.generatePdf,
  generatePdfDataQrCode: () => generatePdfDataQrCode,
  generatePdfQrCode: () => generatePdfQrCode,
  generatePng: () => import_png2.generatePng,
  generatePngDataQrCode: () => generatePngDataQrCode,
  generatePngQrCode: () => generatePngQrCode,
  generateQrCode: () => generateQrCode,
  generateSvg: () => import_svg2.generateSvg,
  generateSvgDataQrCode: () => generateSvgDataQrCode,
  generateSvgPath: () => import_svg2.generateSvgPath,
  generateSvgQrCode: () => generateSvgQrCode
});
module.exports = __toCommonJS(index_exports);
var import_encode = require("./encode.cjs");
var import_pdf = require("./images/pdf.cjs");
var import_png = require("./images/png.cjs");
var import_svg = require("./images/svg.cjs");
var import_matrix = require("./matrix.cjs");
var import_qrcode = require("./qrcode.cjs");
var import_dataurl = require("./utils/dataurl.cjs");
var import_pdf2 = require("./images/pdf.cjs");
var import_png2 = require("./images/png.cjs");
var import_svg2 = require("./images/svg.cjs");
function generateQrCode(message, options) {
  const { ecLevel = "M", url = false } = { ...options };
  const encoded = (0, import_encode.encodeQrCodeMessage)(message, url);
  const qrcode = (0, import_qrcode.generateQrCodeData)(encoded, ecLevel);
  const matrix = (0, import_matrix.generateQrCodeMatrix)(qrcode);
  return {
    version: qrcode.version,
    ecLevel: qrcode.ecLevel,
    size: matrix.length,
    matrix
  };
}
async function generatePngQrCode(message, options) {
  return await (0, import_png.generatePng)(generateQrCode(message, options), options);
}
async function generatePdfQrCode(message, options) {
  return await (0, import_pdf.generatePdf)(generateQrCode(message, options), options);
}
function generateSvgQrCode(message, options) {
  return (0, import_svg.generateSvg)(generateQrCode(message, options), options);
}
async function generatePngDataQrCode(message, options) {
  return (0, import_dataurl.generateDataUrl)(await (0, import_png.generatePng)(generateQrCode(message, options), options), "image/png");
}
async function generatePdfDataQrCode(message, options) {
  return (0, import_dataurl.generateDataUrl)(await (0, import_pdf.generatePdf)(generateQrCode(message, options), options), "application/pdf");
}
function generateSvgDataQrCode(message, options) {
  return (0, import_dataurl.generateDataUrl)((0, import_svg.generateSvg)(generateQrCode(message, options), options), "image/svg+xml");
}
async function generate(message, format, options) {
  switch (format) {
    // plain images
    case "png":
      return await generatePngQrCode(message, options);
    case "pdf":
      return await generatePdfQrCode(message, options);
    case "svg":
      return generateSvgQrCode(message, options);
    // images as data URLs
    case "pngData":
      return await generatePngDataQrCode(message, options);
    case "pdfData":
      return await generatePdfDataQrCode(message, options);
    case "svgData":
      return generateSvgDataQrCode(message, options);
    // coverage ignore next
    default:
      throw new Error(`Unsupported format "${format}"`);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  generate,
  generatePdf,
  generatePdfDataQrCode,
  generatePdfQrCode,
  generatePng,
  generatePngDataQrCode,
  generatePngQrCode,
  generateQrCode,
  generateSvg,
  generateSvgDataQrCode,
  generateSvgPath,
  generateSvgQrCode
});
