// @juit/qrcode/dist/index.mjs
// index.ts
import { encodeQrCodeMessage } from "./encode.mjs";
import { generatePdf } from "./images/pdf.mjs";
import { generatePng } from "./images/png.mjs";
import { generateSvg } from "./images/svg.mjs";
import { generateQrCodeMatrix } from "./matrix.mjs";
import { generateQrCodeData } from "./qrcode.mjs";
import { generateDataUrl } from "./utils/dataurl.mjs";
import { generatePdf as generatePdf2 } from "./images/pdf.mjs";
import { generatePng as generatePng2 } from "./images/png.mjs";
import { generateSvg as generateSvg2, generateSvgPath } from "./images/svg.mjs";
function generateQrCode(message, options) {
  const { ecLevel = "M", url = false } = { ...options };
  const encoded = encodeQrCodeMessage(message, url);
  const qrcode = generateQrCodeData(encoded, ecLevel);
  const matrix = generateQrCodeMatrix(qrcode);
  return {
    version: qrcode.version,
    ecLevel: qrcode.ecLevel,
    size: matrix.length,
    matrix
  };
}
async function generatePngQrCode(message, options) {
  return await generatePng(generateQrCode(message, options), options);
}
async function generatePdfQrCode(message, options) {
  return await generatePdf(generateQrCode(message, options), options);
}
function generateSvgQrCode(message, options) {
  return generateSvg(generateQrCode(message, options), options);
}
async function generatePngDataQrCode(message, options) {
  return generateDataUrl(await generatePng(generateQrCode(message, options), options), "image/png");
}
async function generatePdfDataQrCode(message, options) {
  return generateDataUrl(await generatePdf(generateQrCode(message, options), options), "application/pdf");
}
function generateSvgDataQrCode(message, options) {
  return generateDataUrl(generateSvg(generateQrCode(message, options), options), "image/svg+xml");
}
async function generate(message, format, options) {
  switch (format) {
    // plain images
    case "png":
      return await generatePngQrCode(message, options);
    case "pdf":
      return await generatePdfQrCode(message, options);
    case "svg":
      return generateSvgQrCode(message, options);
    // images as data URLs
    case "pngData":
      return await generatePngDataQrCode(message, options);
    case "pdfData":
      return await generatePdfDataQrCode(message, options);
    case "svgData":
      return generateSvgDataQrCode(message, options);
    // coverage ignore next
    default:
      throw new Error(`Unsupported format "${format}"`);
  }
}
export {
  generate,
  generatePdf2 as generatePdf,
  generatePdfDataQrCode,
  generatePdfQrCode,
  generatePng2 as generatePng,
  generatePngDataQrCode,
  generatePngQrCode,
  generateQrCode,
  generateSvg2 as generateSvg,
  generateSvgDataQrCode,
  generateSvgPath,
  generateSvgQrCode
};
