// @juit/qrcode/dist/matrix.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// matrix.ts
var matrix_exports = {};
__export(matrix_exports, {
  generateQrCodeMatrix: () => generateQrCodeMatrix
});
module.exports = __toCommonJS(matrix_exports);
function init(version) {
  const n = version * 4 + 17;
  const matrix = new Array(n);
  for (let i = 0; i < n; i++) {
    matrix[i] = new Array(n).fill(0);
  }
  return matrix;
}
function fillFinders(matrix) {
  const n = matrix.length;
  for (let i = -3; i <= 3; i++) {
    for (let j = -3; j <= 3; j++) {
      const max = Math.max(i, j);
      const min = Math.min(i, j);
      const pixel = max == 2 && min >= -2 || min == -2 && max <= 2 ? 128 : 129;
      matrix[3 + i][3 + j] = pixel;
      matrix[3 + i][n - 4 + j] = pixel;
      matrix[n - 4 + i][3 + j] = pixel;
    }
  }
  for (let i = 0; i < 8; i++) {
    matrix[7][i] = matrix[i][7] = matrix[7][n - i - 1] = matrix[i][n - 8] = matrix[n - 8][i] = matrix[n - 1 - i][7] = 128;
  }
}
function fillAlignAndTiming(matrix) {
  const n = matrix.length;
  if (n > 21) {
    const len = n - 13;
    let delta = Math.round(len / Math.ceil(len / 28));
    if (delta % 2) delta++;
    const res = [];
    for (let p = len + 6; p > 10; p -= delta) {
      res.unshift(p);
    }
    res.unshift(6);
    for (let i = 0; i < res.length; i++) {
      for (let j = 0; j < res.length; j++) {
        const x = res[i];
        const y = res[j];
        if (matrix[x][y]) continue;
        for (let r = -2; r <= 2; r++) {
          for (let c = -2; c <= 2; c++) {
            const max = Math.max(r, c);
            const min = Math.min(r, c);
            const pixel = max == 1 && min >= -1 || min == -1 && max <= 1 ? 128 : 129;
            matrix[x + r][y + c] = pixel;
          }
        }
      }
    }
  }
  for (let i = 8; i < n - 8; i++) {
    matrix[6][i] = matrix[i][6] = i % 2 ? 128 : 129;
  }
}
function fillStub(matrix) {
  const n = matrix.length;
  for (let i = 0; i < 8; i++) {
    if (i != 6) {
      matrix[8][i] = matrix[i][8] = 128;
    }
    matrix[8][n - 1 - i] = 128;
    matrix[n - 1 - i][8] = 128;
  }
  matrix[8][8] = 128;
  matrix[n - 8][8] = 129;
  if (n < 45) return;
  for (let i = n - 11; i < n - 8; i++) {
    for (let j = 0; j < 6; j++) {
      matrix[i][j] = matrix[j][i] = 128;
    }
  }
}
var fillReserved = (() => {
  const FORMATS = new Array(32);
  const VERSIONS = new Array(40);
  const gf15 = 1335;
  const gf18 = 7973;
  const formatsMask = 21522;
  for (let format = 0; format < 32; format++) {
    let res = format << 10;
    for (let i = 5; i > 0; i--) {
      if (res >>> 9 + i) {
        res = res ^ gf15 << i - 1;
      }
    }
    FORMATS[format] = (res | format << 10) ^ formatsMask;
  }
  for (let version = 7; version <= 40; version++) {
    let res = version << 12;
    for (let i = 6; i > 0; i--) {
      if (res >>> 11 + i) {
        res = res ^ gf18 << i - 1;
      }
    }
    VERSIONS[version] = res | version << 12;
  }
  const EC_LEVELS = { L: 1, M: 0, Q: 3, H: 2 };
  return function fillReserved2(matrix, ecLevel, mask) {
    const N = matrix.length;
    const format = FORMATS[EC_LEVELS[ecLevel] << 3 | mask];
    function _f(k) {
      return format >> k & 1 ? 129 : 128;
    }
    for (let i = 0; i < 8; i++) {
      matrix[8][N - 1 - i] = _f(i);
      if (i < 6) matrix[i][8] = _f(i);
    }
    for (let i = 8; i < 15; i++) {
      matrix[N - 15 + i][8] = _f(i);
      if (i > 8) matrix[8][14 - i] = _f(i);
    }
    matrix[7][8] = _f(6);
    matrix[8][8] = _f(7);
    matrix[8][7] = _f(8);
    const version = VERSIONS[(N - 17) / 4];
    if (!version) return;
    function _v(k) {
      return version >> k & 1 ? 129 : 128;
    }
    for (let i = 0; i < 6; i++) {
      for (let j = 0; j < 3; j++) {
        matrix[N - 11 + j][i] = matrix[i][N - 11 + j] = _v(i * 3 + j);
      }
    }
  };
})();
var fillData = /* @__PURE__ */ (() => {
  const MASK_FUNCTIONS = [
    function(i, j) {
      return (i + j) % 2 == 0;
    },
    function(i, j) {
      void j;
      return i % 2 == 0;
    },
    function(i, j) {
      void i;
      return j % 3 == 0;
    },
    function(i, j) {
      return (i + j) % 3 == 0;
    },
    function(i, j) {
      return (Math.floor(i / 2) + Math.floor(j / 3)) % 2 == 0;
    },
    function(i, j) {
      return i * j % 2 + i * j % 3 == 0;
    },
    function(i, j) {
      return (i * j % 2 + i * j % 3) % 2 == 0;
    },
    function(i, j) {
      return (i * j % 3 + (i + j) % 2) % 2 == 0;
    }
  ];
  return function fillData2(matrix, data, mask) {
    const N = matrix.length;
    let row;
    let col;
    let dir = -1;
    row = col = N - 1;
    const maskFn = MASK_FUNCTIONS[mask];
    let len = data.blockData[data.blockData.length - 1].length;
    for (let i = 0; i < len; i++) {
      for (let b = 0; b < data.blockData.length; b++) {
        if (data.blockData[b].length <= i) continue;
        put(data.blockData[b][i]);
      }
    }
    len = data.ecData[0].length;
    for (let i = 0; i < len; i++) {
      for (let b = 0; b < data.ecData.length; b++) {
        put(data.ecData[b][i]);
      }
    }
    if (col > -1) {
      do {
        matrix[row][col] = maskFn(row, col) ? 1 : 0;
      } while (next());
    }
    function put(byte) {
      for (let mask2 = 128; mask2; mask2 = mask2 >> 1) {
        let pixel = !!(mask2 & byte);
        if (maskFn(row, col)) pixel = !pixel;
        matrix[row][col] = pixel ? 1 : 0;
        next();
      }
    }
    function next() {
      do {
        if (col % 2 ^ (col < 6 ? 1 : 0)) {
          if (dir < 0 && row == 0 || dir > 0 && row == N - 1) {
            col--;
            dir = -dir;
          } else {
            col++;
            row += dir;
          }
        } else {
          col--;
        }
        if (col == 6) {
          col--;
        }
        if (col < 0) {
          return false;
        }
      } while (matrix[row][col] & 240);
      return true;
    }
  };
})();
function calculatePenalty(matrix) {
  const N = matrix.length;
  let penalty = 0;
  for (let i2 = 0; i2 < N; i2++) {
    let pixel = matrix[i2][0] & 1;
    let len = 1;
    for (let j2 = 1; j2 < N; j2++) {
      const p = matrix[i2][j2] & 1;
      if (p == pixel) {
        len++;
        continue;
      }
      if (len >= 5) {
        penalty += len - 2;
      }
      pixel = p;
      len = 1;
    }
    if (len >= 5) {
      penalty += len - 2;
    }
  }
  for (let j2 = 0; j2 < N; j2++) {
    let pixel = matrix[0][j2] & 1;
    let len = 1;
    for (let i2 = 1; i2 < N; i2++) {
      const p = matrix[i2][j2] & 1;
      if (p == pixel) {
        len++;
        continue;
      }
      if (len >= 5) {
        penalty += len - 2;
      }
      pixel = p;
      len = 1;
    }
    if (len >= 5) {
      penalty += len - 2;
    }
  }
  for (let i2 = 0; i2 < N - 1; i2++) {
    for (let j2 = 0; j2 < N - 1; j2++) {
      const s = matrix[i2][j2] + matrix[i2][j2 + 1] + matrix[i2 + 1][j2] + matrix[i2 + 1][j2 + 1] & 7;
      if (s == 0 || s == 4) {
        penalty += 3;
      }
    }
  }
  let i;
  let j;
  function _i(k) {
    return matrix[i][j + k] & 1;
  }
  function _j(k) {
    return matrix[i + k][j] & 1;
  }
  for (i = 0; i < N; i++) {
    for (j = 0; j < N; j++) {
      if (j < N - 6 && _i(0) && !_i(1) && _i(2) && _i(3) && _i(4) && !_i(5) && _i(6)) {
        if (j >= 4 && !(_i(-4) || _i(-3) || _i(-2) || _i(-1))) {
          penalty += 40;
        }
        if (j < N - 10 && !(_i(7) || _i(8) || _i(9) || _i(10))) {
          penalty += 40;
        }
      }
      if (i < N - 6 && _j(0) && !_j(1) && _j(2) && _j(3) && _j(4) && !_j(5) && _j(6)) {
        if (i >= 4 && !(_j(-4) || _j(-3) || _j(-2) || _j(-1))) {
          penalty += 40;
        }
        if (i < N - 10 && !(_j(7) || _j(8) || _j(9) || _j(10))) {
          penalty += 40;
        }
      }
    }
  }
  let numDark = 0;
  for (let i2 = 0; i2 < N; i2++) {
    for (let j2 = 0; j2 < N; j2++) {
      if (matrix[i2][j2] & 1) numDark++;
    }
  }
  penalty += 10 * Math.floor(Math.abs(10 - 20 * numDark / (N * N)));
  return penalty;
}
function generateQrCodeMatrix(code) {
  const matrix = init(code.version);
  fillFinders(matrix);
  fillAlignAndTiming(matrix);
  fillStub(matrix);
  let penalty = Infinity;
  let bestMask = 0;
  for (let mask = 0; mask < 8; mask++) {
    fillData(matrix, code, mask);
    fillReserved(matrix, code.ecLevel, mask);
    const p = calculatePenalty(matrix);
    if (p < penalty) {
      penalty = p;
      bestMask = mask;
    }
  }
  fillData(matrix, code, bestMask);
  fillReserved(matrix, code.ecLevel, bestMask);
  return matrix.map((row) => {
    return row.map((cell) => {
      return !!(cell & 1);
    });
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  generateQrCodeMatrix
});
